filetests/isa/x86/pinsr.clif

```diff
@@ -0,0 +1,23 @@
+test binemit
+; set has_sse41
+target x86_64
+
+function %test_pinsr_i32() {
+ebb0:
+[-, %rax]   v0 = iconst.i32 42
+[-, %rbx]   v1 = iconst.i32 99
+[-, %xmm4]  v2 = splat.i32x4 v0
+; TODO it seems weird that insertlane should return anything, the x86 instruction does not so %xmm7 below is meaningless
```

Comment:
The CLIF definition of `insertlane` looks like:
```
ig.push(
        Inst::new(
            "insertlane",
            r#"
        Insert ``y`` as lane ``Idx`` in x.

        The lane index, ``Idx``, is an immediate value, not an SSA value. It
        must indicate a valid lane index for the type of ``x``.
        "#,
        )
        .operands_in(vec![x, Idx, y])
        .operands_out(vec![a]),
    );
```
Why even have an `out` operand? If it is needed (e.g. other ISAs?), then it would seem like `a` and `x` are supposed to map to the same register--then how do I indicate this in my code?

Reply:
It has an `out` operand because Cranelift IR uses SSA form, and the value after the insert is different from the value before, so we need to represent it with a different `Value`, even if it'll use the same register. To tell the register allocator that the output register is the same as one of the input registers, we can "tie" the output to one of the inputs. See the `"rr"` recipe in cranelift-codegen/meta/src/isa/x86/recipes.rs for example.

Reply:
Ah, I see! `0` is pointing to the first `gpr`?

Reply:
I tried this out for the `insertlane` recipe:
```
        let format = formats.get(f_insert_lane);
        recipes.add_template_recipe(
            EncodingRecipeBuilder::new("r_ib_unsigned_r", f_insert_lane, 2)
                .operands_in(vec![fpr, gpr])
                .operands_out(vec![0])
                .inst_predicate(InstructionPredicate::new_is_unsigned_int(
                    format, "lane", 8, 0,
                ))
                .emit(
                    r#"
                    {{PUT_OP}}(bits, rex2(in_reg1, in_reg0), sink);
                    modrm_rr(in_reg1, in_reg0, sink);
                    let imm:i64 = lane.into();
                    sink.put1(imm as u8);
                "#,
                ),
        );
```
However, when I do this the `filetests/isa/x86/pinsr.clif` test now fails:
```
FAIL filetests/isa/x86/pinsr.clif: binemit(%test_pinsr_b8):
No matching encodings for v3 = insertlane.b8x16 v2, 10, v1 in [Mp3r_ib_unsigned_r#d20]
```
I don't understand why an encoding shouldn't be found. Is there some syntax I'm missing?
```
[-, %rax]   v0 = iconst.i32 42
[-, %rbx]   v1 = iconst.i32 99
[-, %xmm4]  v2 = splat.i32x4 v0
[-, %xmm4]  v3 = insertlane v2, 2, v1    ; bin: 66 0f 3a 22 e3 02
```

Reply:
> Ah, I see! 0 is pointing to the first gpr?

0 is pointing to the first element in the list of input constraints, so in this case, the first is `fpr`.

Reply:
> I don't understand why an encoding shouldn't be found.

There's no clear reason just looking at the code; I'd advise to use a debugger and step in like we did before, to see what goes wrong (either no encodings are found in the encodings list, or the encodings constraints aren't satisfied).

Reply:
Finally fixed this one.